### PR TITLE
OAuth2 improvements

### PR DESCRIPTION
**Pull request by sbrunner**
Be able to have a redirect proxy to be able to share an OAuth2 application on GitHub

The `has_access` is extracted to be able to have some part of the application authenticated on different repository access.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -186,6 +186,12 @@ define the GitHub auth URL (default is `https://api.github.com/user`)
 Using the environment variable `C2C_AUTH_GITHUB_REPO_URL` or the config key `c2c.auth.github.repo_url` to
 define the GitHub auth URL (default is `https://api.github.com/repo`)
 
+Using the environment variable `C2C_AUTH_GITHUB_PROXY_URL` or the config key `c2c.auth.github.auth.proxy_url` to
+define a redirect proxy between GitHub and our application to be able to share an OAuth2 application on GitHub (default is no proxy)
```
Comment (arnaud-morvan) [inline README.md:190]:
```suggestion
define a redirect proxy between GitHub and our application to be able to authenticate users on different GitHub repositories, depending on (I do not really know what, probably something in the request).
```
It desserve something to understand it completely, maybe a link to the proxy repo.

File: c2cwsgiutils/auth.py [hunk 2]
```diff
@@ -178,6 +185,59 @@ def auth_type(settings: Optional[Mapping[str, Any]]) -> Optional[AuthenticationT
     return AuthenticationType.NONE
 
 
+def has_access(
+    request: pyramid.request.Request, repo: Optional[str] = None, access_type: Optional[str] = None
```
Comment (arnaud-morvan) [inline c2cwsgiutils/auth.py:189]: repo and access_type paramters seems never used.

File: README.md [hunk 3]
```diff
@@ -166,7 +166,7 @@ Using the environment variable `C2C_AUTH_GITHUB_CLIENT_SECRET` or the config key
 `c2c.auth.github.client_secret` to define the GitHub application secret (required)
 
 Using the environment variable `C2C_AUTH_GITHUB_SCOPE` or the config key `c2c.auth.github.scope` to define
-the GitHub scope (default is `read:user`)
+the GitHub scope (default is `repo`)
```
Comment (arnaud-morvan) [inline README.md:169]: I do not understand what is a GitHub scope, default is 'repo' but what are other possible values ?

File: CHANGELOG.md [hunk 4]
```diff
@@ -6,6 +6,8 @@
   by `c2c://`).
 - Move back the logging configuration to `production.ini`. It will be read from `gunicorn.conf.py` at startup.
 - Remove the `development.ini` file to simplify the default application template; restore `production.ini` has the default configuration file.
+- When the `auth_type` is GitHub the `is_auth` and the `auth_type` can return different values, the is auth will just check
```
Comment (arnaud-morvan) [inline CHANGELOG.md:9]: I do not understand

> the `is_auth` and the `auth_type` can return different values

As already said : `auth_type` is GitHub
And I suppose `is_auth` is a boolean.

Maybe you want to say : `is_auth` and the `has_access` can return different values

> the is_auth will just check the authentication on GitHub.

Yes, it seems logical as `auth_type` is GitHub, but, as I understand, has_access may do something else.

**Review by arnaud-morvan — COMMENTED** [review]
I do not undestand how we get the repo from the request.

I would like to discuss it by meet before finishing the review.

File: README.md (hunk 3 above)
Comment (sbrunner) [reply README.md]: Just added link to GitHub documentation :-)

File: CHANGELOG.md (hunk 4 above)
Comment (sbrunner) [reply CHANGELOG.md]: updated...

File: README.md (hunk 1 above)
Comment (sbrunner) [reply README.md]: This is not related to `authenticate users on different GitHub repositories` but to `OAuth2 application on GitHub`...
Lint to the proxy added :-)

**Comment by sbrunner** [comment]
+1 for a meet on next week :-)

**Review by arnaud-morvan — APPROVED** [review]

File: c2cwsgiutils/auth.py [hunk 5]
```diff
@@ -178,6 +185,59 @@ def auth_type(settings: Optional[Mapping[str, Any]]) -> Optional[AuthenticationT
     return AuthenticationType.NONE
 
 
+def has_access(
```
Comment (gberaudo) [inline c2cwsgiutils/auth.py:188]: Let's call the function `check_has_access` to avoid confusions.

**Review by gberaudo — APPROVED** [review]

**Pull request merged**